### PR TITLE
FIX #66 #64 #54 #30 #31 #72 #29 #62 - context access - kedro 16.5 - hook auto registration

### LINKED ISSUE #66: Accessing the project context inside hooks implementations
Hi, 

I'm a colleague of @Galileo-Galilei. We discussed the issue of accessing the project template and properties from a plugin. [We raised the point in the kedro project](https://github.com/quantumblacklabs/kedro/issues/506)

What will it solve : 

- No more Template assumptions. We will ask the kedro context where the template component resides (project_paths, conf_paths, src path, ...). That lead to less kedro updates regression, because we will no longer manage a template at our side.

- No more configs loading logic assumptions. We will use the kedro context configLoader and get_crendentials (that can be overrided by user).

That solve : #64 #54 #30 #31 (User will use kedro credentials mechanisms, i'll put details in another issue) and a part of #62

How can we implement it (For now, i see three possibilities):

-  Calling a load_context inside before_pipeline_run hook and setting at the same time a global project_path variable, hoping that the before_pipeline_run hook is always the first hook that will be called in the futur. **--> So hacky**

- Declaring the mlflow hooks as a project context property. Which means that the hooks can easily access to the current context **--> No more auto registration**

- Waiting for the release of the kedro session. It will apparently allows access to the current context **--> We don't know when will it land**

Keep up the good work guys, I'm happy to join you :)

### PR DESCRIPTION
**Pull request by takikadiri**
closes #66 #64 #54 #30 #31 #72 #29 #62

## Description
This PR intend to fix mlflow configs related issues. This indirectly made it easier to upgrade the kedro dependency to 0.16.5 and enable hook's auto-registration, modulo some tests fixing. 

## Development notes

Some important changes : 

* kedro-mlflow can now access to project context properties. That decouple the plugin from the template.
We used this access to change `get_mlflow_config` signature, the test have been updated accordignately.

* `mlflow init` no longer touch the run.py

* I instantiate an Mlflow<type>Hook objects, as .kedro.yml and setuptools entrypoint declared hooks take hook object (not class)

* I add a `.kedro.yml` file, `src` folder and `run.py` file to the kedro project fixture for tests running in kedro 0.16.5

For more details, see changes.

## Checklist

- [x] Read the [contributing](https://github.com/Galileo-Galilei/kedro-mlflow/blob/develop/CONTRIBUTING.md) guidelines
- [x] Open this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Update the documentation to reflect the code changes
- [x] Add a description of this change and add your name to the list of supporting contributions in the [`CHANGELOG.md`](https://github.com/Galileo-Galilei/kedro-mlflow/blob/develop/CHANGELOG.md) file. Please respect [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines.
- [x] Add tests to cover your changes

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

- I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
- I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorised to submit this contribution on behalf of the original creator(s) or their licensees.
- I certify that the use of this contribution as authorised by the Apache 2.0 license does not violate the intellectual property rights of anyone else.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,12 +4,15 @@
 
 ### Added
 
+- kedro-mlflow hooks are now auto-registred (#29)
```
Comment (Galileo-Galilei) [inline CHANGELOG.md:7]: We should not do this in this PR, since it breaks kedro viz (see https://github.com/quantumblacklabs/kedro-viz/issues/260). Can you remove it?

File: CHANGELOG.md [hunk 2]
```diff
@@ -21,6 +24,9 @@
 
 ### Changed
 
+- Bump kedro requirement to <=0.16.5
+- `kedro mlflow init` is no longer touching the run.py file
```
Comment (Galileo-Galilei) [inline CHANGELOG.md:28]: We should likely add a test for the kedro version to ensure retro compatibility, or make it very clear in the docs that if you have kedro ">=0.16,<0.16.4".

File: CHANGELOG.md [hunk 3]
```diff
@@ -21,6 +24,9 @@
 
 ### Changed
 
+- Bump kedro requirement to <=0.16.5
```
Comment (Galileo-Galilei) [inline CHANGELOG.md:27]: We should check that the tests are not broken with kedro<=0.16, because the CI only tests for the last version of kedro, and wa can't be sure that nothing is broken.

File: kedro_mlflow/framework/cli/cli.py [hunk 4]
```diff
@@ -1,16 +1,12 @@
-import os
 import subprocess
 from pathlib import Path
 
 import click
-from kedro import __file__ as KEDRO_PATH
+from kedro.framework.context import get_static_project_data, load_context
```
Comment (Galileo-Galilei) [inline kedro_mlflow/framework/cli/cli.py:5]: I think this breaks retrocompatibilty with kedro<=0.16.4. It should definitely be a separated PR in case we need to rollback. But it's a great idea!

File: kedro_mlflow/framework/cli/cli.py [hunk 5]
```diff
@@ -87,80 +83,86 @@ def init(force, silent):
...
     # mlflow.yml is just a static file,
     # but the name of the experiment is set to be the same as the project
+    conf_root = context.CONF_ROOT
     mlflow_yml = "mlflow.yml"
     write_jinja_template(
         src=TEMPLATE_FOLDER_PATH / mlflow_yml,
         is_cookiecutter=False,
-        dst=project_path / "conf" / "base" / mlflow_yml,
-        python_package=project_globals["python_package"],
+        dst=project_path / conf_root / "base" / mlflow_yml,
```
Comment (Galileo-Galilei) [inline kedro_mlflow/framework/cli/cli.py:97]: I did not remember this was hardcoded, well spotted 👍

File: kedro_mlflow/framework/hooks/node_hook.py [hunk 6]
```diff
@@ -51,6 +90,9 @@ def before_node_run(
         mlflow.log_params(params_inputs)
 
 
+mlflow_node_hooks = MlflowNodeHook()
```
Comment (Galileo-Galilei) [inline kedro_mlflow/framework/hooks/node_hook.py:93]: why is this necessary? Is it for auto registration?

File: kedro_mlflow/framework/hooks/pipeline_hook.py [hunk 7]
```diff
@@ -62,9 +66,13 @@ def before_pipeline_run(
             pipeline: The ``Pipeline`` that will be run.
             catalog: The ``DataCatalog`` to be used during the run.
         """
-        mlflow_conf = get_mlflow_config(
-            project_path=run_params["project_path"], env=run_params["env"]
+        context = load_context(
```
Comment (Galileo-Galilei) [inline kedro_mlflow/framework/hooks/pipeline_hook.py:69]: context -> self.context (in case we need it elsewhere, for now it should have no impact)

File: setup.py [hunk 8]
```diff
@@ -51,10 +43,14 @@ def _parse_requirements(path, encoding="utf-8"):
...
     entry_points={
         "kedro.project_commands": [
-            "kedro_mlflow =  kedro_mlflow.framework.cli.cli:commands"
+            "kedro_mlflow =  kedro_mlflow.framework.cli.cli:commands",
         ],
         "kedro.global_commands": [
-            "kedro_mlflow =  kedro_mlflow.framework.cli.cli:commands"
+            "kedro_mlflow =  kedro_mlflow.framework.cli.cli:commands",
+        ],
+        "kedro.hooks": [
```
Comment (Galileo-Galilei) [inline setup.py:51]: This must be removed because of the kedro-viz bug. We can eventually modify mlflow_init to add what is necessary in the `kedro.yml` file.

File: tests/framework/hooks/test_pipeline_hook.py [hunk 9]
```diff
@@ -4,13 +4,13 @@
 import pytest
 import yaml
 from kedro.extras.datasets.pickle import PickleDataSet
-from kedro.framework.context import KedroContext
+from kedro.framework.context import get_static_project_data
```
Comment (Galileo-Galilei) [inline tests/framework/hooks/test_pipeline_hook.py:7]: I think this only compatible with kedro>=0.16.5

File: tests/mlflow/test_kedro_pipeline_model.py [hunk 10]
```diff
@@ -85,6 +85,9 @@ def test_model_packaging(tmp_path, pipeline_ml_obj):
     )
     assert loaded_model.predict(1) == {"predictions": 2}
 
+    if mlflow.active_run():
+        mlflow.end_run()
+
```
Comment (Galileo-Galilei) [inline tests/mlflow/test_kedro_pipeline_model.py:90]: This is very likely something we can do for all tests. Pytest have a way to call some cleanup code after each tests, and we must force all active mlflow runs (especially when tests fails)

**Review by Galileo-Galilei — COMMENTED** [review]
The PR is really great and adds a lot of new features, but I think we should separate thingsinto at least 3 PR:

- [ ] context access is a whole PR in itself, and tests must be added (for instance, test that mlflow config now works with TemplatedConfigLoader or with an environment different from "local") -> regarding this, all your changes are ok for me and I think we can merge them soon once tests are added.
- [ ] migration to 0.16.5 modifies a lot of things, with low retro compatibility, especially for kedro mlflow init. We can  accept this, but is must be very clear in the documentation(especialy this section: https://github.com/Galileo-Galilei/kedro-mlflow/blob/develop/docs/source/03_tutorial/02_setup.md#automatic-template-update-recommended). I think the switch must be thoroughly thought and made clear before deprecation. For now, we can modify the `kedro.yml` file (or `pyproject.toml`) to register the hook automatically without modifying the `run.py` file.
- [ ] plugin auto registration is a no go while kedro viz bug is not fixed. If people have to chose, I hardly believe they wil choose kedro-mlflow upon kedro-viz 😄 I don't think it is necessary to make the PR right now, but you can have a separated branch and we'll merge it once it goos to go.

File: CHANGELOG.md (hunk 1 above)
Comment (takikadiri) [reply CHANGELOG.md:7]: kedro-viz just work fine with kedro 0.16.5 and auto registred hooks. I'll test it with previous kedro versions

File: CHANGELOG.md (hunk 2 above)
Comment (takikadiri) [reply CHANGELOG.md:28]: yes, we should discuss this particular point in a separate PR.

File: kedro_mlflow/framework/hooks/node_hook.py (hunk 6 above)
Comment (takikadiri) [reply kedro_mlflow/framework/hooks/node_hook.py:93]: It's for registring hooks in .kedro.yml or in setuptools entrypoints.

File: kedro_mlflow/framework/hooks/pipeline_hook.py (hunk 7 above)
Comment (takikadiri) [reply kedro_mlflow/framework/hooks/pipeline_hook.py:69]: Well spoted, it's a `self.context`

File: kedro_mlflow/framework/cli/cli.py (hunk 4 above)
Comment (takikadiri) [reply kedro_mlflow/framework/cli/cli.py:5]: That will definitely break kedro <=0.16.4 compatibility as i just saw that `get_static_project_data` is introduced in kedro 0.16.5. I'll keep the `_get_project_globals` utils

**Comment by Galileo-Galilei** [comment]
Close this issue which has been merged as #91 and #94
